File: scripts/apitypings/testdata/generics/generics.go
package codersdk

import "time"

type Single interface {
	string
}

type Custom interface {
	string | bool | int | time.Duration | []string | *int
}

// Static has all generic fields defined in the field
type Static struct {
	Static Fields[string, int, time.Duration, string] `json:"static"`
}

// Dynamic has some dynamic fields.
type Dynamic[A any, S Single] struct {
	Dynamic    Fields[bool, A, string, S] `json:"dynamic"`
	Comparable bool                       `json:"comparable"`
}

type Complex[C comparable, S Single, T Custom] struct {
	Dynamic    Fields[C, bool, string, S]       `json:"dynamic"`
	Order      FieldsDiffOrder[C, string, S, T] `json:"order"`
	Comparable C                                `json:"comparable"`
	Single     S                                `json:"single"`
	Static     Static                           `json:"static"`
}

type Fields[C comparable, A any, T Custom, S Single] struct {
	Comparable C `json:"comparable"`
	Any        A `json:"any"`

	Custom           T `json:"custom"`
	Again            T `json:"again"`
	SingleConstraint S `json:"single_constraint"`
}

type FieldsDiffOrder[A any, C comparable, S Single, T Custom] struct {
	Fields Fields[C, A, T, S]
}

File: scripts/apitypings/testdata/generics/generics.ts
// From codersdk/generics.go
export interface Complex<C extends comparable, S extends Single, T extends Custom> {
  readonly dynamic: Fields<C, boolean, string, S>
  readonly order: FieldsDiffOrder<C, string, S, T>
  readonly comparable: C
  readonly single: S
  readonly static: Static
}

// From codersdk/generics.go
export interface Dynamic<A extends any, S extends Single> {
  readonly dynamic: Fields<boolean, A, string, S>
  readonly comparable: boolean
}

// From codersdk/generics.go
export interface Fields<C extends comparable, A extends any, T extends Custom, S extends Single> {
  readonly comparable: C
  readonly any: A
  readonly custom: T
  readonly again: T
  readonly single_constraint: S
}

// From codersdk/generics.go
export interface FieldsDiffOrder<A extends any, C extends comparable, S extends Single, T extends Custom> {
  readonly Fields: Fields<C, A, T, S>
}

// From codersdk/generics.go
export interface Static {
  readonly static: Fields<string, number, number, string>
}

// From codersdk/generics.go
export type Custom = string | boolean | number | (readonly string[]) | null

// From codersdk/generics.go
export type Single = string

export type comparable = boolean | number | string | any